src/DynamoCore/Scheduler/UpdateGraphAsyncTask.cs

```diff
@@ -36,6 +37,40 @@ internal UpdateGraphAsyncTask(IScheduler scheduler, bool verboseLogging1) : base
             this.verboseLogging = verboseLogging1;
         }
 
+        /// <summary>
+        /// DFS algorithm to check cyclic dependency in Dynamo graph.
+        /// </summary>
+        /// <param name="node"></param>
+        /// <param name="visitTracker"></param>
+        /// <param name="recursionTracker"></param>
+        /// <param name="cyclicNodes">list of all nodes participating in a cycle</param>
+        /// <returns>true if graph is cyclic, false otherwise</returns>
+        private bool IsGraphCyclic(NodeModel node, HashSet<NodeModel> visitTracker, 
```

Comment:
is there no other similar method already implemented that we could reuse (optionally enhance)?

Reply:
There is a better algorithm used called `StronglyConnectedComponent` but it is used at a lower, `graphnode` level in the engine. We could probably use it for this too at the dynamo `nodemodel` level but we'll need to generalize (templatize) it to work for any graph vertex, which will be tricky as getting the adjacent nodes for each type is very different. It would be easier to rewrite it to make it work specifically with `NodeModel` type.

Reply:
My first goal here was to establish that this idea could work in general, we can always refine the exact algo later and make it more efficient, which at the moment, I can think of the `stronglyconnectedcomponent` algo.

Reply:
Ok
Is this supposed to detect all cycles or just the first one ?

Reply:
I think this will detect only one cycle for a given node but the outer method call will record at most one cycle for **all** nodes and report warnings on all of them.